keps/0009-operator-toolkit.md

```diff
@@ -0,0 +1,296 @@
+- **tasks**: A map of tasks that can be run. These are the atomic runnable unit of a KUDO operator, and are made up of a series of YAML manifests. These are defined more in detail in [Tasks](#tasks).
+- **plans**: A map of plans that can be run. These are the core partitioning unit of a KUDO operator. A plan is intended to run a single "operations task" for an operator, such as backup, restore, deployment, or update. This is defined in detail in [Plans](#plans)
+
+This file undergoes a Go template pass on Instance instantiation before being parsed. This is described more in detail in [Extensions and Bases](#extensions-and-bases)
+
+#### params.yaml
+
+The `params.yaml` file is a struct that defines parameters for operator. This can articulate descriptions, defaults, and triggers etc., In the MySQL example, this looks like:
+
+```yaml
+backupFile:
+  description: "The name of the backup file"
+  default: backup.sql
+password:
+  default: password
+  description: "Password for the mysql instance"
+```
+
+These values are meant to be overridden by Instance resources when instantiating an operator.
+
+This file undergoes a Go template pass on Instance instantiation before being parsed. This is described more in detail in [Parameters](#parameters).
+
+#### templates/
+
+The templates directory contains 0-to-many YAML manifests for operator tasks to use. These are described in more detail in [Templates](#templates).
+
+Each template undergoes a Go template pass for an Instance when it's corresponding Task is run.
+
+### Plans
+
+Plans are the core unit of operation within a KUDO operator. Each KUDO plan represents an individual unit of operations work. This may include backups, restores, deployments, upgrades, compaction, or any other myriad of operations that an application operator may want to perform with KUDO.
+
+A single plan is composed of [Steps](#steps).
+
+### Steps
+
+Steps are used for fine grained partition within a plan. Steps are represented as a list of objects that define the step name and a list of tasks to be run during that step.
+
+All tasks within the same step are applied at the same time. For runnable templates within a task (such as a Deployment, StatefulSet, Pod, etc), parallelism of workloads should be controlled through other Kubernetes primitives such as PodDisruptionBudget, and added to the task where the relevant workload is run.
+
+If a step contains a runnable task that has readiness probes defined, the step waits until the readiness probe succeeds before moving onto the next step.
+
+### Tasks
+
+Tasks are a map of task name to a list of templates that should be executed in that step. With [Extensions and Bases](#extensions-and-bases), tasks can be represented in multiple forms, as long as the end result is a ready-to-run YAML manifest for a Kubernetes object.
+
+#### Files
+
+If a filename is specified, KUDO will execute a Go Template on the relevant filename, described more in detail in [Templates](#templates).
```

Comment:
I like the idea of using Go templating. Would we have additional Framework logic, e.g. `kubectl kudo kafka create topic` or so live as an extension in those Files or how would that be different from Extensions and Bases?

Reply:
One of our examples had a `plans` folder to eventually support this use case but won't be scoped in this work, so we're thinking this is a different KEP. Right now, something like that would run an existing plan (say a create-topic plan).